src/Compilers/CSharp/Portable/Symbols/NativeIntegerTypeSymbol.cs

```diff
@@ -93,50 +162,249 @@ internal NativeIntegerTypeSymbol(NamedTypeSymbol underlyingType) : base(underlyi
+        internal override bool Equals(TypeSymbol? other, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverrideOpt = null)
+        {
+            if (other is null)
+            {
+                return false;
+            }
+            if ((object)this == other)
+            {
+                return true;
+            }
+            if (!_underlyingType.Equals(other, comparison, isValueTypeOverrideOpt))
+            {
+                return false;
+            }
+            return (comparison & TypeCompareKind.IgnoreNativeIntegers) != 0 ||
+                other.IsNativeIntegerType;
+        }
 
         public override int GetHashCode() => _underlyingType.GetHashCode();
 
         void Cci.IReference.Dispatch(Cci.MetadataVisitor visitor)
         {
-            // NativeIntegerTypeSymbol should not be used in emit.
+            // Emit should use underlying symbol only.
             throw ExceptionUtilities.Unreachable;
         }
 
         private ImmutableArray<NamedTypeSymbol> GetInterfaces(ConsList<TypeSymbol>? basesBeingResolved)
         {
             if (_lazyInterfaces.IsDefault)
             {
-                ImmutableInterlocked.InterlockedCompareExchange(ref _lazyInterfaces, makeInterfaces(_underlyingType.InterfacesNoUseSiteDiagnostics(basesBeingResolved)), default(ImmutableArray<NamedTypeSymbol>));
+                var interfaces = _underlyingType.InterfacesNoUseSiteDiagnostics(basesBeingResolved).SelectAsArray((Func<NamedTypeSymbol, NativeIntegerTypeMap, NamedTypeSymbol>)((type, map) => map.SubstituteNamedType(type)), GetTypeMap());
+                ImmutableInterlocked.InterlockedInitialize(ref _lazyInterfaces, interfaces);
             }
             return _lazyInterfaces;
+        }
+
+        private NativeIntegerTypeMap GetTypeMap()
+        {
+            if (_lazyTypeMap is null)
+            {
+                Interlocked.CompareExchange(ref _lazyTypeMap, new NativeIntegerTypeMap(this), null);
+            }
+            return _lazyTypeMap;
+        }
+
+        /// <summary>
+        /// Replaces references to underlying type with references to native integer type.
+        /// </summary>
+        internal TypeWithAnnotations SubstituteUnderlyingType(TypeWithAnnotations type) => type.SubstituteType(GetTypeMap());
+
+        /// <summary>
+        /// Replaces references to underlying type with references to native integer type.
+        /// </summary>
+        internal NamedTypeSymbol SubstituteUnderlyingType(NamedTypeSymbol type) => GetTypeMap().SubstituteNamedType(type);
+
+        /// <summary>
+        /// Replaces references to underlying type with references to native integer type.
+        /// </summary>
+        internal ImmutableArray<CustomModifier> SubstituteCustomModifiers(ImmutableArray<CustomModifier> customModifiers) => GetTypeMap().SubstituteCustomModifiers(customModifiers);
```

Comment:
>SubstituteCustomModifiers [](start = 48, length = 25)

I'll remove this. It doesn't seem necessary to update custom modifiers since there is no consumer.